hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java

```diff
@@ -0,0 +1,93 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hdds.utils;
+
+
+import org.apache.ozone.test.tag.Native;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.mockito.MockedStatic;
+import org.mockito.Mockito;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.apache.hadoop.hdds.utils.NativeConstants.ROCKS_TOOLS_NATIVE_LIBRARY_NAME;
+import static org.apache.hadoop.hdds.utils.NativeLibraryLoader.NATIVE_LIB_TMP_DIR;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.ArgumentMatchers.same;
+
+/**
+ * Test class for NativeLibraryLoader.
+ */
+public class TestNativeLibraryLoader {
+
+  private static Stream<String> nativeLibraryDirectoryLocations()
+      throws IOException {
+    return Stream.of("", File.createTempFile("prefix", "suffix")
+        .getParentFile().getAbsolutePath(), null);
+  }
+
+  @Native(ROCKS_TOOLS_NATIVE_LIBRARY_NAME)
+  @ParameterizedTest
+  @MethodSource("nativeLibraryDirectoryLocations")
+  public void testNativeLibraryLoader(
+      String nativeLibraryDirectoryLocation) {
+    Map<String, Boolean> libraryLoadedMap = new HashMap<>();
+    NativeLibraryLoader loader = new NativeLibraryLoader(libraryLoadedMap);
+    try (MockedStatic<NativeLibraryLoader> mockedNativeLibraryLoader =
+             Mockito.mockStatic(NativeLibraryLoader.class,
+                 Mockito.CALLS_REAL_METHODS)) {
+      mockedNativeLibraryLoader.when(() ->
+              NativeLibraryLoader.getSystemProperty(same(NATIVE_LIB_TMP_DIR)))
+          .thenReturn(nativeLibraryDirectoryLocation);
+      mockedNativeLibraryLoader.when(() -> NativeLibraryLoader.getInstance())
+          .thenReturn(loader);
+      Assertions.assertTrue(NativeLibraryLoader.getInstance()
+          .loadLibrary(ROCKS_TOOLS_NATIVE_LIBRARY_NAME));
+      Assertions.assertTrue(NativeLibraryLoader
+          .isLibraryLoaded(ROCKS_TOOLS_NATIVE_LIBRARY_NAME));
+      // Mocking to force copy random bytes to create a lib file to
+      // nativeLibraryDirectoryLocation. But load library will fail.
+      mockedNativeLibraryLoader.when(() ->
+          NativeLibraryLoader.getResourceStream(anyString()))
+          .thenReturn(new ByteArrayInputStream(new byte[]{0, 1, 2, 3}));
+      String dummyLibraryName = "dummy_lib";
+      NativeLibraryLoader.getInstance().loadLibrary(dummyLibraryName);
+      NativeLibraryLoader.isLibraryLoaded(dummyLibraryName);
+      // Checking if the resource with random was copied to a temp file.
+      File[] libPath =
+          new File(nativeLibraryDirectoryLocation == null ? "" :
```

Comment:
It brings back to the same point if we really need to test for null. You can simply have two test cases
Case 1: empty string: `""`,
Case 2: non-empty string: `File.createTempFile("prefix", "suffix").getParentFile().getAbsolutePath()`


This is test not source. In my opinion fall back to empty string in case of null doesn't make sense in test.

Reply:
by default this will go to /tmp we don't want that. We want it to copy to the current working directory.

Reply:
We are checking if the native library file gets copied to the cwd in case of null.

Reply:
Ah, I see `nativeLibraryDirectoryLocation` is returned by `mockedNativeLibraryLoader`.

It was my miss. I didn't see `nativeLibraryDirectoryLocation` was not used other than this place.